File: src/main/java/ru/mm/surv/capture/service/impl/FfmpegRestartableStream.java
package ru.mm.surv.capture.service.impl;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.ObjectFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Service;
import ru.mm.surv.capture.service.FfmpegStream;

import javax.annotation.PreDestroy;
import java.io.File;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;

@Service
@Primary
@Slf4j
public class FfmpegRestartableStream implements FfmpegStream {

    private final ObjectFactory<FfmpegMultiStream> ffmpegMultiStreamFactory;
    private FfmpegMultiStream ffmpegMultiStream;

    @Autowired
    public FfmpegRestartableStream(ObjectFactory<FfmpegMultiStream> ffmpegMultiStreamFactory) {
        this.ffmpegMultiStreamFactory = ffmpegMultiStreamFactory;
    }

    @Override
    public synchronized void start() {
        if (!isActive()) {
            ffmpegMultiStream = ffmpegMultiStreamFactory.getObject();
        }
        ffmpegMultiStream.start();
    }

    @Override
    @PreDestroy
    public synchronized void stop() {
        if (isActive()) {
            ffmpegMultiStream.stop();
            ffmpegMultiStream = null;
        }
    }

    @Override
    public boolean isActive() {
        if (ffmpegMultiStream != null && !ffmpegMultiStream.isActive()) {
            ffmpegMultiStream = null;
        }
        return ffmpegMultiStream != null;
    }

    @Override
    public Collection<String> getStreamNames() {
        return Optional
                .ofNullable(ffmpegMultiStream)
                .map(FfmpegStream::getStreamNames)
                .orElse(Collections.emptyList());
    }

    @Override
    public Optional<File> getThumb(String stream) {
        return Optional
                .ofNullable(ffmpegMultiStream)
                .flatMap(e -> e.getThumb(stream));
    }
}

File: src/main/java/ru/mm/surv/capture/service/impl/FfmpegRestartableStream.kt
package ru.mm.surv.capture.service.impl

import org.springframework.context.annotation.Primary
import org.springframework.beans.factory.ObjectFactory
import org.springframework.stereotype.Service
import ru.mm.surv.capture.service.FfmpegStream
import java.io.File
import javax.annotation.PreDestroy

@Service
@Primary
class FfmpegRestartableStream(
    private val ffmpegMultiStreamFactory: ObjectFactory<FfmpegMultiStream>) : FfmpegStream {
    private var ffmpegMultiStream: FfmpegMultiStream? = null

    @Synchronized
    override fun start() {
        if (!isActive()) {
            ffmpegMultiStream = ffmpegMultiStreamFactory.getObject()
        }
        ffmpegMultiStream!!.start()
    }

    @PreDestroy
    @Synchronized
    override fun stop() {
        if (isActive()) {
            ffmpegMultiStream!!.stop()
            ffmpegMultiStream = null
        }
    }

    override fun isActive(): Boolean {
        if (ffmpegMultiStream != null && !ffmpegMultiStream!!.isActive()) {
            ffmpegMultiStream = null
        }
        return ffmpegMultiStream != null
    }

    override fun streamNames(): Collection<String> {
        return ffmpegMultiStream
            ?.streamNames()
            ?: emptyList()
    }

    override fun getThumb(stream: String): File? {
        return ffmpegMultiStream?.getThumb(stream);
    }
}
